strawberry/tools/depth_limit_validator.py

```diff
@@ -0,0 +1,171 @@
+    InlineFragmentNode,
+    OperationDefinitionNode,
+)
+from graphql.validation import ValidationContext, ValidationRule
+
+
+def depth_limit_validator(max_depth: int, options=None, callback=None):
+    """
+    Creates a validator for the GraphQL query depth
+
+    This is a Python port of https://github.com/stems/graphql-depth-limit
+
+    - max_depth - The maximum allowed depth for any operation in a GraphQL document.
+    - options
+        - options.ignore - Stops recursive depth checking based on a field name.
+        Either a string or regexp to match the name, or a function that returns
+        a boolean.
+    - callback - Called each time validation runs. Receives an Object which is a
+    map of the depths for each operation.
+    """
+    if not options:
+        options = {}
+
+    class DepthLimitValidator(ValidationRule):
+        def __init__(self, validation_context: ValidationContext):
+            document = validation_context.document
+            definitions = document.definitions
+
+            fragments = get_fragments(definitions)
+            queries = get_queries_and_mutations(definitions)
+            query_depths = {}
+
+            for name in queries:
+                query_depths[name] = determine_depth(
+                    queries[name],
+                    fragments,
+                    0,
+                    max_depth,
+                    validation_context,
+                    name,
+                    options,
+                )
+
+            if callable(callback):
+                callback(query_depths)
+            super().__init__(validation_context)
+
+    return DepthLimitValidator
+
+
+def get_fragments(definitions: List[DefinitionNode]):
+    fragments = {}
+    for definition in definitions:
+        if isinstance(definition, FragmentDefinitionNode):
+            fragments[definition.name.value] = definition
+
+    return fragments
+
+
+# This will actually get both queries and mutations.
+# We can basically treat those the same
+def get_queries_and_mutations(definitions: List[DefinitionNode]):
+    operations = {}
+
+    for definition in definitions:
+        if isinstance(definition, OperationDefinitionNode):
+            operations[
+                definition.name.value if definition.name else "anonymous"
+            ] = definition
+
+    return operations
+
+
+introspection_regex = re.compile(r"^__")
+
+
+def determine_depth(
+    node, fragments, depth_so_far, max_depth, context, operation_name, options
+) -> int:
+    if depth_so_far > max_depth:
+        context.report_error(
+            GraphQLError(
+                f"'{operation_name}' exceeds maximum operation depth of {max_depth}",
+                [node],
+            )
+        )
+        return depth_so_far
+
+    if isinstance(node, FieldNode):
+        # by default, ignore the introspection fields which begin with double underscores
+        should_ignore = introspection_regex.match(node.name.value) or see_if_ignored(
```

Comment:
We could use `node.name.value.startswith("__")` instead of the regex, which may be slightly faster and maybe easier to read. (I know a regex is used because its a port :) )

Reply:
There is also this little documented helper function which could be reused if we want to https://github.com/strawberry-graphql/strawberry/blob/e6b005ce98df53f923c477fc11ab69c0a8f58736/strawberry/extensions/tracing/utils.py#L8

Reply:
Ah good spot, will change.